In [2]:
import pandas as pd

# Import datasets
df_prods = pd.read_csv("/Users/boukaskasbrahim/Desktop/Achievement4_Project/Original_Data/products.csv")
df_ords = pd.read_csv("/Users/boukaskasbrahim/Desktop/Achievement4_Project/Prepared Data/orders_wrangled.csv")

# Quick checks
print(df_prods.shape)
print(df_ords.shape)

df_prods.head()
df_ords.head()


(49693, 5)
(3421083, 6)


order_id  user_id  order_number  order_dow  order_hour_of_day  \
0   2539329        1             1          2                  8   
1   2398795        1             2          3                  7   
2    473747        1             3          3                 12   
3   2254736        1             4          4                  7   
4    431534        1             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [3]:
df_ords.describe()


order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

### Data Consistency Check – df_ords (describe output)

After running `df.describe()` on the `df_ords` dataframe, here are the main observations:

- order_id, user_id, order_number:  
  Values range from 1 up to the expected maximums. Nothing looks unusual here.  

- order_dow:  
  Minimum = 0 and Maximum = 6 → correct, since there are 7 days (0–6).  

- order_hour_of_day:  
  Minimum = 0 and Maximum = 23 → correct, since hours of the day range from 0–23.  

- days_since_prior_order:  
  Minimum = 0 and Maximum = 30 → looks correct, since 0 means “same day” and 30 means “a long gap since the last order.”  

Conclusion:  
All minimum and maximum values fall within expected ranges. No anomalies detected in this first consistency check.


In [5]:
for col in df_ords.columns.tolist():
    weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis=1)
    if len(df_ords[weird]) > 0:
        print(col)


/var/folders/fh/z04k5mcd3xz6ftdmhk_5j1m00000gn/T/ipykernel_21103/2365901188.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis=1)
/var/folders/fh/z04k5mcd3xz6ftdmhk_5j1m00000gn/T/ipykernel_21103/2365901188.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis=1)
/var/folders/fh/z04k5mcd3xz6ftdmhk_5j1m00000gn/T/ipykernel_21103/2365901188.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis=1)
/var/folders/fh/z04k5mcd3xz6ftdmhk_5j1m00000gn/T/ipykernel_21103/2365901188.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].a

### Mixed-Type Data Check – df_ords

I ran a loop to check each column for mixed data types using `map()`.  

- Result: No mixed-type columns were found.  
- Action: No changes were needed.  

**Conclusion:** All columns in the `df_ords` dataframe contain consistent data types.


In [6]:
df_ords.isnull().sum()


order_id                       0
user_id                        0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

### Missing Values – df_ords

I checked for missing values with `isnull().sum()`.  

- All columns are complete except **`days_since_prior_order`**, which has 206,209 missing values.  
- This is normal, because when a customer makes their very first order, there is no “previous order” to calculate days from.  

**Action:** I decided to keep these missing values as they are, since they carry meaning (they show first orders).  


In [7]:
df_dups = df_ords[df_ords.duplicated()]
print(df_dups.shape)


(0, 6)


no duplicate rows.

In [10]:

df_prods.to_csv("/Users/boukaskasbrahim/Desktop/Achievement4_Project/Prepared Data/products_checked.csv", index=False)

df_ords.to_csv("/Users/boukaskasbrahim/Desktop/Achievement4_Project/Prepared Data/orders_checked.csv", index=False)



I exported the final cleaned dataframes:  
- `products_checked.csv`  
- `orders_checked.csv`  
